shelley/chain-and-ledger/executable-spec/test/Generator/Update.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Generator.Update
+  ( genPParams )
+  where
+
+import           Data.Ratio ((%))
+import           Hedgehog (Gen)
+
+import qualified Hedgehog.Gen as Gen
+
+import           BaseTypes (Nonce (NeutralNonce), UnitInterval, mkNonce)
+import           Coin (Coin (..))
+import           Examples (unsafeMkUnitInterval)
+import           Generator.Core (genNatural)
+import qualified Hedgehog.Range as Range
+import           Numeric.Natural (Natural)
+import           PParams (PParams (..))
+import           Slot (Epoch (Epoch))
+
+
+genRationalInThousands :: Integer -> Integer -> Gen Rational
+genRationalInThousands lower upper =
+  (% 1000) <$>
+    Gen.integral (Range.linear lower upper)
+
+genIntervalInThousands :: Integer -> Integer -> Gen UnitInterval
+genIntervalInThousands lower upper =
+  unsafeMkUnitInterval <$> genRationalInThousands lower upper
+
+-- TODO @uroboros for now, keeping minA/B at zero until we generate fees in genTx
+genPParams :: Gen PParams
+genPParams = mkPParams <$> pure 0 -- _minfeeA
+                       <*> pure 0 -- _minfeeB
+                       <*> szGen  -- (maxBBSize, maxBHSize, maxTxSize)
+                       -- keyDeposit
+                       <*> (Coin <$> Gen.integral (Range.linear 0 50))
+                       -- keyMinRefund: 0.1-0.5
+                       <*> genIntervalInThousands 100 500
+                       -- keyDecayRate: 0.001-0.1
+                       <*> genRationalInThousands 1 100
+                       -- poolDeposit
+                       <*> Gen.integral (Range.linear 0 500)
+                       -- poolMinRefund: 0.1-0.7
+                       <*> genIntervalInThousands 100 700
+                       -- poolDecayRate: 0.001-0.1
+                       <*> genRationalInThousands 1 100
+                       -- eMax
+                       <*> (Epoch <$> Gen.integral (Range.linear 20 500))
+                       -- nOpt
+                       <*> Gen.integral (Range.linear 1 100)
+                       -- a0: 0.01-1.0
+                       <*> genRationalInThousands 10 1000
+                       -- rho: 0.001-0.009
+                       <*> genIntervalInThousands 1 9
+                       -- tau: 0.1-0.3
+                       <*> genIntervalInThousands 100 300
+                       -- activeSlotCoeff: 0-1
+                       <*> genIntervalInThousands 0 1000
```

Comment:
in the case where we really want to see / test `0`, it might be better to use an exponential distribution or some weighted choice

Reply:
👍